### PR TITLE
ci: refactor msbuild workflow

### PR DESCRIPTION
**Pull request by ReenigneArcher**
This PR makes some changes the msbuild.yml workflow. I am happy to describe any changes in further detail if needed.

- Add additional events for the workflow
  - On `pull_request` allows you to verify if PRs have successful builds and/or download the artifacts for testing. This is limited to opened, synchronized (when new commits are pushed to the PR branch), and re-opened events.
  - On `push` is limited to only the `master` branch.
  - On `workflow_dispatch` allows you to run this workflow manually if you ever need to check the build state.
- The `concurrency` changes will ensure only a single workflow will run for each ref. This means if a PR contributor pushes new changes to their branch, it will cancel the in process workflow and trigger a new run.
- `env` section was shortened, but equivalent to the previous
- Updated versions for three of the github actions used
- Removed default values such as:
  - `shell: pwsh` as this is the default for Windows runners
  - `working-directory: ${{env.GITHUB_WORKSPACE}}` as this is the default for all job steps
- Used line continuations for long commands
- Added a single line break between steps to improve readability

There is a repository setting to configure how github workflows run from outside collaborators. The options are:
  - require approval for first time contributors
  - require approval for all outside collaborates (This is what I use for LizardByte, as it reduces runner consumption when I am not ready to merge PRs)

No matter which option you select, PRs from forks will never have access to secrets. There are none in your workflow currently, so this is just FYI.

P.S. In another PR, I would like to add dependabot to help you keep actions and git submodules updated automatically. Unfortunately, it does not have support for vcpkg (see: https://github.com/dependabot/dependabot-core/issues/2027) 

## Summary by CodeRabbit

- **New Features**
  - Enhanced workflow configuration to trigger on pull requests, improving the development process.

- **Improvements**
  - Updated action versions for better performance and compatibility.
  - Streamlined script executions for improved readability and maintainability.
  - Introduced concurrency settings to manage in-progress workflows effectively.
  - Added security enhancements for handling certificates in the build process.

### CODE REVIEW

File: .github/workflows/msbuild.yml [hunk 1]
```diff
@@ -1,74 +1,79 @@
...
-      working-directory: ${{env.GITHUB_WORKSPACE}}
       run: .\generate-thirdparty-projects.bat
+
     - name: Restore NuGet
-      working-directory: ${{env.GITHUB_WORKSPACE}}
       run: nuget restore ${{env.SOLUTION_FILE_PATH}}
+
     - name: Load Certificate
+      env:
+        CERTIFICATE_FILE: ${{ secrets.CERTIFICATE_FILE }}
```
Comment (ReenigneArcher) [inline .github/workflows/msbuild.yml:54]: I missed that there was a secret used in the workflow.

Can we load a dummy certificate when it's a PR?

Comment (ReenigneArcher) [reply .github/workflows/msbuild.yml:54]: @TheElixZammuto any thoughts on this? My assumption is the certificate is only needed for production builds that will be sent to the MSFT Store?

Comment (TheElixZammuto) [reply .github/workflows/msbuild.yml:54]: It's counterintuitive, but MS Store does not use the certificate inside secrets.CERTIFICATE_FILE, unlike App Store and Google Play. This is needed for install on Xboxes, and it's just a self signed certificate in pfx format

Comment (ReenigneArcher) [reply .github/workflows/msbuild.yml:54]: Okay, so for PRs from forks, we could generate a certificate?

And for pushes or PRs from this repo, use the secret?

Comment (ReenigneArcher) [reply .github/workflows/msbuild.yml:92]: I gave this an attempt, but the build is failing to validate the certificate.

```
        "D:\a\moonlight-xbox\moonlight-xbox\moonlight-xbox-dx.sln" (default target) (1) ->
       "D:\a\moonlight-xbox\moonlight-xbox\moonlight-xbox-dx.vcxproj" (default target) (2) ->
       (_ValidateSigningCertificate target) -> 
         C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VisualStudio\v17.0\AppxPackage\Microsoft.AppXPackage.Targets(894,5): error : Certificate could not be opened: cert.pfx. [D:\a\moonlight-xbox\moonlight-xbox\moonlight-xbox-dx.vcxproj]
       C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VisualStudio\v17.0\AppxPackage\Microsoft.AppXPackage.Targets(894,5): error : The specified network password is not correct. [D:\a\moonlight-xbox\moonlight-xbox\moonlight-xbox-dx.vcxproj]
       C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VisualStudio\v17.0\AppxPackage\Microsoft.AppXPackage.Targets(894,5): error :  [D:\a\moonlight-xbox\moonlight-xbox\moonlight-xbox-dx.vcxproj]

    3211 Warning(s)
    1 Error(s)
```

I'm not totally sure how the validation occurs, but it seems that it's trying to make a request to the address.

Most of what I tried is from here: https://stackoverflow.com/questions/11403333

**Comment by ReenigneArcher** [comment]
@TheElixZammuto this is working now. https://github.com/ReenigneArcher/moonlight-xbox/actions/runs/10231641854

**Comment by ReenigneArcher** [comment]
After this is merged, I plan on working on a way to cache the third party builds as that part is quite time consuming (~25 minutes).

**Comment by ReenigneArcher** [comment]
@TheElixZammuto re-running the original job won't work as it uses the original workflow without my new fixes.

Do you have a button like this?

image

**Comment by ReenigneArcher** [comment]
Looks like i need to tweak something else. I have it working in my repo, so should be close.

https://github.com/ReenigneArcher/moonlight-xbox/pull/1

**Comment by ReenigneArcher** [comment]
I think I've fixed this.

**Comment by TheElixZammuto** [comment]
Thank you as always!

**Pull request merged**